lib/licence_location_identifier.rb

```diff
@@ -5,27 +5,44 @@ class LicenceLocationIdentifier < LocationIdentifier
   # to find the most appropriate council from the "providing tier"
   # if not, then return the closest authority available.
   def self.find_slug(areas, artefact = nil)
-    councils = areas.map do |area|
-      [area["type"], area["codes"] && area["codes"]["govuk_slug"]]
-    end
+    new(areas, artefact).find_slug
+  end
 
-    if artefact and providing_tier = artefact["details"].try(:[], "licence").try(:[], "local_service").try(:[], "providing_tier")
-      councils_by_tier = Hash[councils.collect do |k, v|
-        [self.identify_tier(k), v]
-      end]
+  attr_reader :areas, :artefact
 
-      appropriate_authority = providing_tier.map do |tier|
-        councils_by_tier[tier]
-      end.compact.first
+  def initialize(areas, artefact)
+    @areas = areas
+    @artefact = artefact
+  end
 
-      return appropriate_authority unless appropriate_authority.nil?
+  def find_slug
+    if providing_tier && matching_authority_by_tier
+      matching_authority_by_tier.try(:[], "codes").try(:[], "govuk_slug")
+    elsif matching_authority_by_area_type
+      matching_authority_by_area_type.try(:[], "codes").try(:[], "govuk_slug")
     end
+  end
+
+private
 
-    authorities = Hash[councils]
-    self.authority_types.each do |type|
-      return authorities[type] unless authorities[type].nil?
+  def providing_tier
+    return unless artefact
+
+    artefact["details"].try(:[], "licence").try(:[], "local_service").try(:[], "providing_tier")
+  end
+
+  def matching_authority_by_tier
+    @_authority_by_tier ||= areas.detect do |area|
+      providing_tier.include?(LocationIdentifier.identify_tier(area["type"]))
```

Comment:
Given:
```ruby
providing_tier: ['county', 'unitary']
areas: [{type: 'unitary', name: 'foo'}, {type: 'county', name: 'bar'}]
```
In the old implementation we'd get "bar" for `matching_authority_by_tier`, but in the new implementation we'd get "foo".  When the service is provided by the "all" tier, which maps to `["district", "unitary", "county"]` for providing_tier (see: https://github.com/alphagov/publisher/blob/master/lib/local_service_importer.rb#L27-L36) then this would mean we might get the "county" instead of the "district" if it appears first in the list from mapit.  This might be bad.